app/domain/authentication/authn_azure/validate_application_identity.rb

```diff
@@ -0,0 +1,102 @@
+    Err = Errors::Authentication::AuthnAzure
+    # Possible Errors Raised: RoleNotFound, InvalidApplicationIdentity
+
+    ValidateApplicationIdentity = CommandClass.new(
+      dependencies: {
+        role_class:                 ::Role,
+        resource_class:             ::Resource,
+        application_identity_class: ApplicationIdentity,
+        logger:                     Rails.logger
+      },
+      inputs:       %i(account service_id username xms_mirid_token_field oid_token_field)
+    ) do
+
+      def call
+        token_identity_from_claims
+        # compare xms_mirid with what is defined in annotations
+        validate_application_identity
+      end
+
+      private
+
+      # xms_mirid is a term in Azure to define a claim that describes the resource that holds the encoding of the instance's
+      # among other details the subscription_id, resource group, and provider identity needed for authorization.
+      # xms_mirid is one of the fields in the JWT token. This function will extract the relevant information from
+      # xms_mirid claim and populate a representative hash with the appropriate fields.
+      def token_identity_from_claims
+        raise Err::ClaimInInvalidFormat unless validate_xms_mirid_format
+
+        @token_identity = {
+          subscription_id: xms_mirid_hash["subscriptions"],
+          resource_group:  xms_mirid_hash["resourcegroups"]
+        }
+
+        # determines which Azure assigned identity is provided in annotations
+        # user-assigned-identity:
+        #   - validates that the correct provider (Microsoft.ManagedIdentity) for a user identity is defined in the
+        #     claim. If so, the 'user_assigned_identity' attribute will be added to the hash with the corresponding
+        #     value from xms_mirid claim
+        # system-assigned-identity:
+        #   - validates that the correct provider (Microsoft.Compute) for a system identity is defined in the
+        #     claim and its resource is one that we support. At current, we only support a virtualMachines resource.
+        #     If these conditions are met, a 'system_assigned_identity' attribute will be added to the hash with its
+        #     value being the oid field in the JWT token.
+        @logger.debug(Log::ExtractingIdentityForAuthentication.new("#{xms_mirid_hash["providers"]}/#{xms_mirid_hash.keys[-1]}"))
+        if xms_mirid_hash["providers"] == "Microsoft.ManagedIdentity"
+          @token_identity[:user_assigned_identity] = xms_mirid_hash["userAssignedIdentities"]
+          @token_identity[:resource_name]          = xms_mirid_hash["userAssignedIdentities"]
+        else
+          @token_identity[:system_assigned_identity] = @oid_token_field
+          @token_identity[:resource_name]            = @oid_token_field
+        end
+      end
+
+      def validate_xms_mirid_format
+        xms_mirid_hash.length == 4 && xms_mirid_hash.key?("subscriptions" && "resourcegroups" && "providers")
+      end
```

Comment:
This should be `valid_xms_mirid_format?` as it's a boolean method. Or even better:
``` 
def validate_xms_mirid_format
  valid_xms_mirid_format = xms_mirid_hash.length == 4 && xms_mirid_hash.key?("subscriptions" && "resourcegroups" && "providers")
  raise Err::ClaimInInvalidFormat unless valid_format
end
```

And then call this method instead of what you have above.